### PR TITLE
BATCHIMP-27 - Update mode doesn't allow to set empty a specific field

### PR DESCRIPTION
**Pull request by slauriere**
This pull request reverts the previous implementation of the BATCHIMP-27 feature and provides a new implementation based on an additional parameter "honor empty values" in the configuration as specified in [BATCHIMP-27 comment-102400](https://jira.xwiki.org/browse/BATCHIMP-27?focusedCommentId=102400&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-102400). Whether the empty values are an empty string or null, the value is used as is (respectively empty string or null) to set the corresponding object property value, when the "honor empty values" option is set to true. When the option is set to false, empty or null values have no effect on the existing data.

### CODE REVIEW

File: xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java [hunk 1]
```diff
@@ -1348,7 +1348,7 @@ public Document marshalDataToDocumentObjects(DocumentReference pageName, Map<Str
                 if (!key.startsWith("doc.")) {
                     PropertyInterface prop = defaultClass.get(key);
 
-                    if (value != null && (!StringUtils.isEmpty(stringValue) || overwrite == Overwrite.UPDATE_EVEN_WHEN_EMPTY_INPUT_CELL)) {
+                    if (!StringUtils.isEmpty(stringValue) || config.getHonorEmptyValues()) {
                         boolean addtotags = false;
```
Comment (lucaa) [inline xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java:1350]: The TODO from line 1346 above needs to be removed / updated to reflect the new behaviour.

File: xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java [hunk 2]
```diff
@@ -1358,7 +1358,13 @@ public Document marshalDataToDocumentObjects(DocumentReference pageName, Map<Str
                             // make this check, although value should really be instanceof List at this point since it
                             // was properly prepared by the parseAndValidatePageData function
                             if (addtotags && value instanceof List) {
-                                tagList.addAll((List) value);
+                                for (Object item: (List)value) {
+                                    // Never add an empty tag whatever is the parameter value 'honor empty values' as it
+                                    // makes no sense to add an empty tag
+                                    if (item != null && StringUtils.isEmpty(item.toString())) {
+                                        tagList.add(item.toString());
```
Comment (lucaa) [inline xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java]: 1/ I'm not sure why this code was changed in this commit, as it's not related to any actual usage of the honor empty values
2/ if anything, the value list should be checked for being empty or not , not each individual value from the parsed list
3/ the test seems wrong, I'd say it needs to test for not empty rather than empty. 

For all these reasons, maybe the handling of the tags could be skipped for this fix...

File: xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java [hunk 3]
```diff
@@ -1369,44 +1375,55 @@ public Document marshalDataToDocumentObjects(DocumentReference pageName, Map<Str
...
 
             // set tags, only if needed.
-            // TODO: fix this test here, it should depend on an "empty overwrites" parameter, which should say whether
-            // an empty value is considered significant or not, which should also apply to properties: for now we
-            // overwrite it even with an empty tagList, if the fields for tags is set to something and that something is
-            // void, maybe we shouldn't
-            if (fieldsfortags != null && fieldsfortags.size() > 0) {
+            // NB: we consider that the tags have to be replaced, not added to possibly existing ones, just like
+            // what happens with the other multi-valued properties.
+            if (tagList.size() > 0) {
```
Comment (lucaa) [inline xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java]: I'm not sure I am comfortable with the idea of changing this test here. You can add an extra condition to the original condition to enter this code, but I would like to avoid that all of a sudden (for any reason) tagList may end-up non empty and tag objects start to be added to the documents...
If we want to change the condition and the way the tags are handled, I'd say we'd do that in a different commit for a different issue, this issue is about deciding if empty values from the source file are ignored or not...

File: xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java (hunk 1 above)
Comment (slauriere) [reply xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java:1350]: Right, I removed it in the latest commit.

File: xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java (hunk 2 above)
Comment (slauriere) [reply xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java]: Right, I removed this code, which was out of scope and buggy indeed, apologies.

File: xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java (hunk 3 above)
Comment (slauriere) [reply xwiki-batchimport-api/src/main/java/org/xwiki/batchimport/internal/DefaultBatchImport.java]: I removed this code as well.
